application/client.core/src/app/environment/components/sidebar/shell/environment/environment_variables/component.ts

```diff
@@ -0,0 +1,198 @@
+                        env.editing.value = true;
+                        this._prevValue = env.value;
+                    } else {
+                        env.editing.variable = true;
+                        this._prevVariable = env.variable;
+                    }
+                }
+            },
+            {
+                caption: 'Remove',
+                handler: () => {
+                    this._changeFocus();
+                    this.information.env = this.information.env.filter((ngEnv: IEnvironment) => {
+                        return ngEnv.variable !== env.variable && ngEnv.value !== env.value;
+                    });
+                },
+            },
+        ];
+        ContextMenuService.show({
+            items: items,
+            x: event.pageX,
+            y: event.pageY,
+        });
+    }
+
+    public _ng_onKeyUp(event: KeyboardEvent) {
+        if (event.key !== 'Enter') {
+            return;
+        }
+        this._finishCreating(true);
+        this._setCustomStatus();
+        this._sortInformation();
+    }
+
+    public _ng_onClick(env: IEnvironment) {
+        if (this._ng_selected !== env) {
+            this._changeFocus();
+        }
+        if (this._ng_selected !== undefined) {
+            this._ng_selected.selected = false;
+        }
+        this._ng_selected = env;
+        this._ng_selected.selected = true;
+    }
+
+    public _ng_onDoubleClick(env: IEnvironment, type: string) {
+        this._changeFocus(env);
+        if (type === this._ng_value) {
+            env.editing.value = true;
+            this._prevValue = env.value;
+        } else {
+            env.editing.variable = true;
+            this._prevVariable = env.variable;
+        }
+    }
+
+    public _ng_add() {
+        this._ng_selected = undefined;
+        this._ng_adding = true;
+        if (this._ng_variableAdd !== undefined) {
+            this._ng_variableAdd.nativeElement.scrollIntoView();
+        }
+    }
+
+    public _ng_remove() {
+        if (this._ng_selected !== undefined) {
+            this.information.env = this.information.env.filter((env: IEnvironment) => {
+                return (env.value !== this._ng_selected.value && env.variable !== this._ng_selected.variable);
+            });
+        }
+    }
+
+    public _ng_onCancel() {
+        this._finishCreating(false);
+        this.close();
+    }
+
+    private _changeFocus(newEditEnv?: IEnvironment) {
+        if (this._editing !== undefined) {
+            this._dropEditChanges();
+            this._editing.editing.value = false;
+            this._editing.editing.variable = false;
+        }
+        this._finishCreating(false);
+        this._editing = newEditEnv;
+    }
+
+    private _dropEditChanges() {
+        if (this._editing.editing.value) {
+            this._editing.value = this._prevValue;
+        } else if (this._editing.editing.variable) {
+            this._editing.variable = this._prevVariable;
+        }
```

Comment:
Declaration 
```
private _editing: IEnvironment | undefined;
```
`_editing` can be `undefined`
In this case here we have potential exception `Cannot read value of undefined`
```
       if (this._editing.editing.value) {                      <-- here
            this._editing.value = this._prevValue;
        } else if (this._editing.editing.variable) {       <-- here
            this._editing.variable = this._prevVariable;
        }
```

Reply:
The marked lines 146-150 are part of the method `_dropEditChanges()`, which is only called inside `_changeFocus(...)`.
The method `_changeFocus(...)` checks in it's first line `if (this._editing !== undefined)` and only calls `_dropEditChanges()` after making sure that it's not `undefined`.

```    private _changeFocus(newEditEnv?: IEnvironment) {
        if (this._editing !== undefined) {            // Making sure it's not undefined
            this._dropEditChanges();                    // Calling the method below
            this._editing.editing.value = false;
            this._editing.editing.variable = false;
        }
        this._finishCreating(false);
        this._editing = newEditEnv;
    }

    private _dropEditChanges() {                              // Since we checked if this._editing is undefined, we don't need to check here anymore
        if (this._editing.editing.value) {
            this._editing.value = this._prevValue;
        } else if (this._editing.editing.variable) {
            this._editing.variable = this._prevVariable;
        }
    }```

Reply:
If method "this._dropEditChanges()" called only from "_changeFocus()" would it be better move checks and rest stuff there?
```
// body of _dropEditChanges
        if (this._editing !== undefined) {           
            if (this._editing.editing.value) {                      
                this._editing.value = this._prevValue;
            } else if (this._editing.editing.variable) {       
                this._editing.variable = this._prevVariable;
            }                 
            this._editing.editing.value = false;
            this._editing.editing.variable = false;
        }

// In this case in _changeFocus, you just call it
        this._dropEditChanges();
        this._finishCreating(false);
        this._editing = newEditEnv;

```

Reply:
yeah, I've changed it a bit more than that, so I think the problem should also be resolved now
Needed some changes for the other comments as well, that's why